src/db.rs

```diff
@@ -0,0 +1,113 @@
+
+use hmac::{crypto_mac::InvalidKeyLength, Hmac, Mac};
+use redis::{Client as RedisClient, Commands, RedisError};
+use sha2::Sha256;
+
+use app_errors::{AppError, AppErrorKind, AppResult};
+use settings::Settings;
+
+/// Database client.
+///
+/// Really just a thin wrapper
+/// around `redis::Client`,
+/// with some logic for generating keys via HMAC
+/// as a safeguard against leaking PII.
+#[derive(Debug)]
+pub struct Client {
+    client: RedisClient,
+    hmac_key: String,
+}
+
+impl Client {
+    /// Instantiate a db client.
+    pub fn new(settings: &Settings) -> Self {
+        Self {
+            client: RedisClient::open(
+                format!("redis://{}:{}/", settings.redis.host, settings.redis.port).as_str(),
+            ).expect("redis connection error"),
+            hmac_key: settings.hmackey.clone(),
+        }
+    }
+
+    /// Read data.
+    pub fn get(&self, key: &str, data_type: DataType) -> AppResult<String> {
+        let key = self.generate_key(key, data_type)?;
+        self.client.get(key.as_str()).map_err(From::from)
+    }
+
+    /// Read and delete data.
+    pub fn consume(&self, key: &str, data_type: DataType) -> AppResult<String> {
+        let key = self.generate_key(key, data_type)?;
+        let key_str = key.as_str();
+        self.client
+            .get(key_str)
+            .map(|metadata| {
+                self.client.del::<&str, u8>(key_str).ok();
+                metadata
+            }).map_err(From::from)
+    }
+
+    /// Store data.
+    ///
+    /// Any data previously stored for the key
+    /// will be clobbered.
+    pub fn set(&self, key: &str, data: &str, data_type: DataType) -> AppResult<()> {
+        let key = self.generate_key(key, data_type)?;
+        self.client.set(key.as_str(), data).map_err(From::from)
+    }
+
+    fn generate_key(&self, key: &str, data_type: DataType) -> AppResult<String> {
+        let mut hmac: Hmac<Sha256> = Hmac::new_varkey(self.hmac_key.as_bytes())?;
+        hmac.input(key.as_bytes());
+        Ok(format!("{}:{:x}", data_type, hmac.result().code()))
+    }
+}
+
+/// Date types included in this store.
+#[derive(Clone, Copy, Debug)]
+pub enum DataType {
+    Configuration,
+    DeliveryProblem,
+    MessageData,
+}
+
+impl Display for DataType {
+    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
+        write!(
+            formatter,
+            "{}",
+            match *self {
+                DataType::Configuration => "cfg",
+                DataType::DeliveryProblem => "del",
+                DataType::MessageData => "msg",
+            }
+        )
+    }
+}
+
+impl From<RedisError> for AppError {
+    fn from(error: RedisError) -> AppError {
+        AppErrorKind::DbError(format!("redis error: {:?}", error)).into()
+    }
+}
+
+impl From<InvalidKeyLength> for AppError {
+    fn from(error: InvalidKeyLength) -> AppError {
+        AppErrorKind::DbError(format!("hmac key error: {:?}", error)).into()
```

Comment:
I wonder if we should create a specific error for this one, like `AppErrorKind::HmacKeyError`, and remove the prefixes (`redis error: ` and `hmac key error: `) from these two. The advantage of that is a clearer error message and a specific errno for each. 

This is not necessarily something I think should be changed, when creating the errors for this app I was never sure about the best way to do this, what do you think, @philbooth ?

Reply:
> I was never sure about the best way to do this

Me neither!

> what do you think, @philbooth ?

I'll make the change!